### PR TITLE
Bonsai tries: update code to store location and account

### PR DESCRIPTION
**Pull request by matkt**
Signed-off-by: Karim TAAM <karim.t2am@gmail.com>

## PR description

Update code to store location and account :
- isMainAccountTrie (boolean) : to know if we are in the main account trie vs if we are going down an account storage trie.
-  location (Bytes) : location in the trie

## Changelog

- [ ] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/worldstate/AccountTrieNodeDataRequest.java [hunk 1]
```diff
@@ -28,23 +29,28 @@
 
 class AccountTrieNodeDataRequest extends TrieNodeDataRequest {
 
-  AccountTrieNodeDataRequest(final Hash hash) {
-    super(RequestType.ACCOUNT_TRIE_NODE, hash);
+  private final boolean isMainAccountTrie;
```
Comment (shemnon) [inline ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/worldstate/AccountTrieNodeDataRequest.java:32]: Do we need `isMainAccountTrie`?  There is only one set of accounts, and the account storage is wrapped up in the StorageTrieNode requests.

Comment (matkt) [reply ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/worldstate/AccountTrieNodeDataRequest.java:32]: So if I understand correctly, to know if we are in the main account trie vs if we are going down an account storage trie we just have to look at the type of the request ?

main account trie -> AccountTrieNodeDataRequest
account storage trie -> StorageTrieNodeDataRequest

And if it's true this information is already present and it is the RequestType field

**Review by shemnon — COMMENTED** [review]
I'm trying to figure out where the location gets updated with each descent.  You need to look at branch nodes (16 possible children) and extension nodes (1 child, but a variable amount of path data)

**Comment by matkt** [comment]
> I'm trying to figure out where the location gets updated with each descent. You need to look at branch nodes (16 possible children) and extension nodes (1 child, but a variable amount of path data)

For me it was done here https://github.com/hyperledger/besu/blob/5241747ba4a62e08953b9584ae03732196b69795/ethereum/trie/src/main/java/org/hyperledger/besu/ethereum/trie/StoredNodeFactory.java#L174 and here https://github.com/hyperledger/besu/blob/5241747ba4a62e08953b9584ae03732196b69795/ethereum/trie/src/main/java/org/hyperledger/besu/ethereum/trie/StoredNodeFactory.java#L195 but maybe I missed something

**Comment by shemnon** [comment]
Will this PR just be making the data available or will it become the fast sync PR?  i.e. do you want it approved now?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateKeyValueStorage.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+import org.hyperledger.besu.plugin.services.storage.KeyValueStorageTransaction;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Predicate;
+
+public class BonsaiWorldStateKeyValueStorage implements WorldStateStorage {
+
+  private static final byte[] WORLD_ROOT_HASH_KEY = "worldRoot".getBytes(StandardCharsets.UTF_8);
```
Comment (shemnon) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateKeyValueStorage.java:35]: public static final.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateKeyValueStorage.java [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+  private static final byte[] WORLD_BLOCK_HASH_KEY = "worldBlockHash".getBytes(StandardCharsets.UTF_8);
```
Comment (shemnon) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateKeyValueStorage.java:37]: public static final

**Comment by matkt** [comment]
> Will this PR just be making the data available or will it become the fast sync PR? i.e. do you want it approved now?

It's not completely fastsync PR. this is mostly the first BonsaiTrieWorldStateStorage version and the addition of information in the NodeDataRequest. Why both at the same time because I found it difficult to do one without the other

**Comment by matkt** [comment]
> > Will this PR just be making the data available or will it become the fast sync PR? i.e. do you want it approved now?
> 
> It's not completely fastsync PR. this is mostly the first BonsaiTrieWorldStateStorage version and the addition of information in the NodeDataRequest. Why both at the same time because I found it difficult to do one without the other

I just moved the code to another PR https://github.com/hyperledger/besu/pull/1765 . I'll close that one

**Pull request closed without merging**